### PR TITLE
Fix num options

### PR DESCRIPTION
**Pull request by sven-oly**
This updates lots of things, including a start on loading data via JSON fetch. It also restructures the directory structure, creating subdirectories with versions under testResults/node and testResults/rust.

├── node
│   ├── v16.19.1
│   │   ├── coll_shift_short
│   │   │   ├── coll_test_shift.html
│   │   │   ├── coll_test_shift.json
│   │   │   ├── failing_tests
│   │   │   │   └── failing_tests.json
│   │   │   ├── pass
│   │   │   │   └── pass.json
│   │   │   ├── test_errors
│   │   │   │   └── test_errors.json
│   │   │   └── unsupported
│   │   │       └── unsupported.json
│   │   ├── lang_names
│   │   │   ├── failing_tests
│   │   │   │   └── failing_tests.json
│   │   │   ├── lang_name_test_file.html
│   │   │   ├── lang_name_test_file.json
│   │   │   ├── pass
│   │   │   │   └── pass.json
│   │   │   ├── test_errors
│   │   │   │   └── test_errors.json
│   │   │   └── unsupported
│   │   │       └── unsupported.json
│   │   └── number_fmt
...

### CODE REVIEW

**Comment by echeran** [comment]
Also merge in the latest from upstream main to pull in #60 which fixes the CI problems for the end-to-end check.

File: testdriver/testplan.py [hunk 1]
```diff
@@ -119,6 +130,14 @@ def requestExecutorInfo(self):
       try:
         self.jsonOutput["platform"] = json.loads(result)
         self.platformVersion =  self.jsonOutput["platform"]["platformVersion"]
+        self.icuVersion =  self.jsonOutput["platform"]["icuVersion"]
+
+        # Reset the output patch based on the version.
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply testdriver/testplan.py:135]: OK

File: testdriver/ddtargs.py [hunk 2]
```diff
@@ -121,6 +121,15 @@ def setCommonArgs(parser):
   parser.add_argument('--parallel_mode', default=None)
   parser.add_argument('--run_limit', default=None)
 
+  # Arguments for setting versions of executors
+  parser.add_argument('--node_version', default='node')  # Sets the version of node to test
+  parser.add_argument('--rust_version', default='latest')
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply testdriver/ddtargs.py:126]: OK

File: testdriver/datasets.py [hunk 3]
```diff
@@ -172,15 +172,17 @@ class ICU4XVersion(Enum):
...
-    "16.1.0": "69.1",
-    "15.14.0": "68.1",
-    "13.14.0": "66.1",
+    "18": "72.1",
+    "16": "71.1",
+    "14": "70.1",
     }
 
+# For Rust
+RustICUVersionMap = {
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply testdriver/datasets.py:181]: Sure

File: testdriver/testplan.py [hunk 4]
```diff
@@ -351,13 +370,26 @@ def processBatchOfTests(self, testsToSend):
...
     for item in result.split('\n'):
       if self.debug > 1:
         print(' RESULT %d = (%d)  >%s<' % (index, len(item), item))
       if item and len(item) > 0:
         # Check for debug data
         if item[0] == "#":
           print('#### DEBUG OUTPUT = %s' % item)
+
+          # Process some types of errors
+          if item[1:3] == "!!":
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply testdriver/testplan.py:383]: Good suggestion. Updating with comments.

File: testgen/testdata_gen.py [hunk 5]
```diff
@@ -193,6 +188,9 @@ def generateNumberFmtTestDataObjects(rawtestdata, count=0):
 
   all_tests_list = []
   verify_list = []
+  expected_count = len(test_list) * len([3, 7, 11]) * len(numbers_to_test) + count
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply testgen/testdata_gen.py:191]: Done. This whole file needs to be restructured using classes.

File: testgen/testdata_gen.py [hunk 6]
```diff
@@ -206,23 +204,20 @@ def generateNumberFmtTestDataObjects(rawtestdata, count=0):
     for l in { 3, 7, 11 }:
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply testgen/testdata_gen.py:204]: Good suggestion. I also refactored "t" and "n" in this function.

File: testgen/testdata_gen.py [hunk 7]
```diff
@@ -193,6 +188,9 @@ def generateNumberFmtTestDataObjects(rawtestdata, count=0):
 
   all_tests_list = []
   verify_list = []
+  expected_count = len(test_list) * len([3, 7, 11]) * len(numbers_to_test) + count
+  max_digits = math.ceil(math.log10(expected_count + 1))
+  print('  Expected count = %s' % expected_count)
   for t in test_list:
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply testgen/testdata_gen.py:194]: Done

File: testgen/testdata_gen.py [hunk 8]
```diff
@@ -206,23 +204,20 @@ def generateNumberFmtTestDataObjects(rawtestdata, count=0):
     for l in { 3, 7, 11 }:
       for n in range(len(numbers_to_test)):
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply testgen/testdata_gen.py:205]: Done

File: testgen/testdata_gen.py [hunk 9]
```diff
@@ -317,14 +313,16 @@ def generateCollTestDataObjects(testdata_list):
       n += 1
 
   print('Coll Test: %d lines processed' % n)
-
+  # Compute max size needed for label number
+  num_samples = len(colltestdata)
+  max_digits = math.ceil(math.log10(num_samples + 1))
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply testgen/testdata_gen.py:318]: Done

File: testgen/testdata_gen.py [hunk 10]
```diff
@@ -258,23 +244,31 @@ def resolveOptions(raw_options, skeleton_list):
...
 def generateDcmlFmtTestDataObjects(rawtestdata, count=0):
+  original_count = count
   recommentline = re.compile('^\s*#')
   test_list = rawtestdata.splitlines()
 
   all_tests_list = []
   verify_list = []
+
+  expected = len(test_list) + count
+  print('  expected count = %s' % (len(test_list) -1))
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply testgen/testdata_gen.py:261]: Thx. Replacing print with logging in this file.

File: verifier/testreport.py [hunk 11]
```diff
@@ -214,15 +215,48 @@ def save_report(self):
...
+      try:
+        os.mkdir(dir_name)
+      except FileExistsError:
+        # It's OK if the directory is already there.
+        pass
+      try:
+        os.mkdir(category_dir_name)
+      except FileExistsError:
+        # It's OK if the directory is already there.
+        pass
```
[the remark this answers is not in the document]
Comment (sven-oly) [reply verifier/testreport.py:248]: Done

**Review by echeran — APPROVED** [review]

**Pull request merged**